# HG changeset patch
# User igerasim
# Date 1490938197 25200
#      Thu Mar 30 22:29:57 2017 -0700
# Node ID 079cd6c5de272ba5d61b79669a0522d5c0be1178
# Parent  56e0ab47dbec09db094eb302ffaa7116d8223de9
8176760: Better handling of PKCS8 material
Reviewed-by: ascarpino, ahgross

--- jdk/src/share/classes/sun/security/pkcs/PKCS8Key.java
+++ jdk/src/share/classes/sun/security/pkcs/PKCS8Key.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 1996, 2013, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1996, 2017, Oracle and/or its affiliates. All rights reserved.
  * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
  *
  * This code is free software; you can redistribute it and/or modify it
@@ -32,6 +32,7 @@ import java.security.Key;
 import java.security.KeyRep;
 import java.security.PrivateKey;
 import java.security.KeyFactory;
+import java.security.MessageDigest;
 import java.security.Security;
 import java.security.Provider;
 import java.security.InvalidKeyException;
@@ -420,18 +421,9 @@ public class PKCS8Key implements Private
             // that encoding
             byte[] b2 = ((Key)object).getEncoded();
 
-            // do the comparison
-            int i;
-            if (b1.length != b2.length)
-                return false;
-            for (i = 0; i < b1.length; i++) {
-                if (b1[i] != b2[i]) {
-                    return false;
-                }
-            }
-            return true;
+            // time-constant comparison
+            return MessageDigest.isEqual(b1, b2);
         }
-
         return false;
     }
 
